File: packages/agent/src/context/llm-summarizer.ts
/**
 * @fileoverview LLM-Based Summarizer
 *
 * Spawns a Haiku subagent to generate intelligent summaries for context
 * compaction. Falls back to KeywordSummarizer on any failure.
 *
 * Follows the same subsession pattern as WebFetch (agent-factory.ts)
 * and LedgerWriter (ledger-writer.ts).
 */

import type { Message } from '@core/types/index.js';
import { KeywordSummarizer, type Summarizer, type SummaryResult, type ExtractedData } from './summarizer.js';
import { COMPACTION_SUMMARIZER_PROMPT } from './system-prompts/compaction-summarizer.js';
import { createLogger } from '@infrastructure/logging/index.js';
import {
  SUMMARIZER_MAX_SERIALIZED_CHARS,
  SUMMARIZER_ASSISTANT_TEXT_LIMIT,
  SUMMARIZER_THINKING_TEXT_LIMIT,
  SUMMARIZER_TOOL_RESULT_TEXT_LIMIT,
  SUMMARIZER_SUBAGENT_TIMEOUT_MS,
} from './constants.js';
import { SUBAGENT_MODEL } from '@llm/providers/model-ids.js';

const logger = createLogger('context:llm-summarizer');

// =============================================================================
// Types
// =============================================================================

export interface LLMSummarizerDeps {
  spawnSubsession: (params: {
    task: string;
    model: string;
    systemPrompt: string;
    toolDenials: { denyAll: true };
    maxTurns: number;
    blocking: true;
    timeout: number;
  }) => Promise<{
    success: boolean;
    output?: string;
    error?: string;
  }>;
}

// =============================================================================
// LLMSummarizer
// =============================================================================

export class LLMSummarizer implements Summarizer {
  private deps: LLMSummarizerDeps;
  private fallback: KeywordSummarizer;

  constructor(deps: LLMSummarizerDeps) {
    this.deps = deps;
    this.fallback = new KeywordSummarizer();
  }

  async summarize(messages: Message[]): Promise<SummaryResult> {
    try {
      const serialized = serializeMessages(messages);

      const result = await this.deps.spawnSubsession({
        task: serialized,
        model: SUBAGENT_MODEL,
        systemPrompt: COMPACTION_SUMMARIZER_PROMPT,
        toolDenials: { denyAll: true },
        maxTurns: 1,
        blocking: true,
        timeout: SUMMARIZER_SUBAGENT_TIMEOUT_MS,
      });

      if (!result.success || !result.output) {
        logger.warn('LLM summarizer subagent failed, using fallback', {
          error: result.error,
          hasOutput: !!result.output,
        });
        return this.fallback.summarize(messages);
      }

      const parsed = parseResponse(result.output);
      if (!parsed) {
        logger.warn('LLM summarizer returned invalid JSON, using fallback', {
          outputLength: result.output.length,
          outputPreview: result.output.slice(0, 200),
        });
        return this.fallback.summarize(messages);
      }

      logger.info('LLM summarizer produced summary', {
        narrativeLength: parsed.narrative.length,
        hasExtractedData: !!parsed.extractedData,
        filesModified: parsed.extractedData?.filesModified?.length ?? 0,
        keyDecisions: parsed.extractedData?.keyDecisions?.length ?? 0,
      });

      return parsed;
    } catch (error) {
      logger.warn('LLM summarizer threw, using fallback', {
        error: (error as Error).message,
      });
      return this.fallback.summarize(messages);
    }
  }
}

// =============================================================================
// Message Serialization
// =============================================================================

/**
 * Serialize messages to a text transcript for the summarizer subagent.
 * Caps output at SUMMARIZER_MAX_SERIALIZED_CHARS — if exceeded, keeps first 25% + last 25%
 * with a middle omission marker.
 */
export function serializeMessages(messages: Message[]): string {
  const lines: string[] = [];

  for (const msg of messages) {
    if (msg.role === 'user') {
      const text = typeof msg.content === 'string'
        ? msg.content
        : msg.content
            .filter((b): b is { type: 'text'; text: string } => b.type === 'text' && 'text' in b)
            .map(b => b.text)
            .join('\n');
      lines.push(`[USER] ${text}`);
    } else if (msg.role === 'assistant') {
      for (const block of msg.content) {
        if (block.type === 'text') {
          const text = (block as { text: string }).text;
          lines.push(`[ASSISTANT] ${text.slice(0, SUMMARIZER_ASSISTANT_TEXT_LIMIT)}`);
        } else if (block.type === 'thinking') {
          const thinking = (block as { thinking?: string }).thinking ?? '';
          if (thinking) {
            lines.push(`[THINKING] ${thinking.slice(0, SUMMARIZER_THINKING_TEXT_LIMIT)}`);
          }
        } else if (block.type === 'tool_use') {
          const tc = block as { name: string; arguments: Record<string, unknown> };
          const keyArgs = extractKeyArgs(tc.arguments);
          lines.push(`[TOOL_CALL] ${tc.name}(${keyArgs})`);
        }
      }
    } else if (msg.role === 'toolResult') {
      const text = typeof msg.content === 'string'
        ? msg.content
        : msg.content
            .filter((b): b is { type: 'text'; text: string } => b.type === 'text' && 'text' in b)
            .map(b => b.text)
            .join('\n');
      const prefix = msg.isError ? '[TOOL_ERROR]' : '[TOOL_RESULT]';
      lines.push(`${prefix} ${text.slice(0, SUMMARIZER_TOOL_RESULT_TEXT_LIMIT)}`);
    }
  }

  const full = lines.join('\n');

  if (full.length <= SUMMARIZER_MAX_SERIALIZED_CHARS) {
    return full;
  }

  const quarter = Math.floor(SUMMARIZER_MAX_SERIALIZED_CHARS / 4);
  const head = full.slice(0, quarter);
  const tail = full.slice(-quarter);
  return `${head}\n\n[... ${full.length - quarter * 2} characters omitted ...]\n\n${tail}`;
}

/**
 * Extract key arguments from tool call arguments for compact display.
 */
function extractKeyArgs(args: Record<string, unknown>): string {
  const keys = ['file_path', 'path', 'command', 'pattern', 'url', 'query'];
  const parts: string[] = [];
  for (const key of keys) {
    if (key in args && typeof args[key] === 'string') {
      const val = args[key] as string;
      parts.push(`${key}: ${val.slice(0, 100)}`);
    }
  }
  return parts.join(', ');
}

// =============================================================================
// Response Parsing
// =============================================================================

/**
 * Parse the subagent's JSON response. Strips markdown fences if present.
 * Returns null on any parse failure.
 */
function parseResponse(raw: string): SummaryResult | null {
  let text = raw.trim();

  // Strip markdown code fences
  const fenceMatch = text.match(/^```(?:json)?\s*\n?([\s\S]*?)\n?\s*```$/);
  if (fenceMatch) {
    text = fenceMatch[1]!.trim();
  }

  let parsed: Record<string, unknown>;
  try {
    parsed = JSON.parse(text);
  } catch {
    return null;
  }

  if (typeof parsed.narrative !== 'string' || !parsed.narrative) {
    return null;
  }

  const extractedData = normalizeExtractedData(
    (parsed.extractedData as Record<string, unknown>) ?? {}
  );

  return {
    narrative: parsed.narrative,
    extractedData,
  };
}

/**
 * Normalize extracted data, defaulting missing fields to empty values.
 */
function normalizeExtractedData(raw: Record<string, unknown>): ExtractedData {
  return {
    currentGoal: typeof raw.currentGoal === 'string' ? raw.currentGoal : '',
    completedSteps: asStringArray(raw.completedSteps),
    pendingTasks: asStringArray(raw.pendingTasks),
    keyDecisions: asDecisionArray(raw.keyDecisions),
    filesModified: asStringArray(raw.filesModified),
    topicsDiscussed: asStringArray(raw.topicsDiscussed),
    userPreferences: asStringArray(raw.userPreferences),
    importantContext: asStringArray(raw.importantContext),
    thinkingInsights: asStringArray(raw.thinkingInsights),
  };
}

function asStringArray(val: unknown): string[] {
  if (!Array.isArray(val)) return [];
  return val.filter((v): v is string => typeof v === 'string');
}

function asDecisionArray(val: unknown): Array<{ decision: string; reason: string }> {
  if (!Array.isArray(val)) return [];
  return val.filter(
    (v): v is { decision: string; reason: string } =>
      typeof v === 'object' &&
      v !== null &&
      typeof (v as Record<string, unknown>).decision === 'string' &&
      typeof (v as Record<string, unknown>).reason === 'string'
  );
}

File: packages/agent-rs/crates/tron-runtime/src/context/llm_summarizer.rs
//! LLM-based summarizer.
//!
//! Spawns a subagent (Haiku) for intelligent context summarization.
//! Falls back to [`KeywordSummarizer`] on any failure.
//!
//! Uses [`serialize_messages`](crate::summarizer::serialize_messages) and
//! [`parse_summarizer_response`](crate::summarizer::parse_summarizer_response)
//! from the shared summarizer module.

use async_trait::async_trait;
use tracing::warn;

use tron_core::messages::Message;

use super::summarizer::{
    parse_summarizer_response, serialize_messages, KeywordSummarizer, Summarizer,
};
use super::types::SummaryResult;

// =============================================================================
// Types
// =============================================================================

/// Result of spawning a summarizer subsession.
#[derive(Clone, Debug)]
pub struct SubsessionResult {
    /// Whether the subsession succeeded.
    pub success: bool,
    /// Output text from the subsession.
    pub output: Option<String>,
    /// Error message if failed.
    pub error: Option<String>,
}

/// Dependency for spawning a summarizer subsession.
///
/// Abstracted as a trait for testability.
#[async_trait]
pub trait SubsessionSpawner: Send + Sync {
    /// Spawn a summarizer subsession with the given task transcript.
    async fn spawn_summarizer(&self, task: &str) -> SubsessionResult;
}

// =============================================================================
// LlmSummarizer
// =============================================================================

/// LLM-based summarizer that spawns a subagent for intelligent summaries.
///
/// On any failure (subsession error, invalid JSON, missing narrative),
/// falls back to [`KeywordSummarizer`].
pub struct LlmSummarizer<S: SubsessionSpawner> {
    spawner: S,
    fallback: KeywordSummarizer,
}

impl<S: SubsessionSpawner> LlmSummarizer<S> {
    /// Create a new LLM summarizer with the given subsession spawner.
    pub fn new(spawner: S) -> Self {
        Self {
            spawner,
            fallback: KeywordSummarizer,
        }
    }
}

#[async_trait]
impl<S: SubsessionSpawner> Summarizer for LlmSummarizer<S> {
    async fn summarize(
        &self,
        messages: &[Message],
    ) -> Result<SummaryResult, Box<dyn std::error::Error + Send + Sync>> {
        let serialized = serialize_messages(messages);

        let result = self.spawner.spawn_summarizer(&serialized).await;

        let output = match result {
            ref r if r.success => r.output.clone(),
            _ => None,
        };
        let Some(output) = output else {
            warn!(
                error = result.error.as_deref().unwrap_or("unknown"),
                "LLM summarizer subagent failed, using fallback"
            );
            return self.fallback.summarize(messages).await;
        };
        if let Ok(parsed) = parse_summarizer_response(&output) {
            Ok(parsed)
        } else {
            warn!(
                output_length = output.len(),
                "LLM summarizer returned invalid JSON, using fallback"
            );
            self.fallback.summarize(messages).await
        }
    }
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use tron_core::content::AssistantContent;
    use tron_core::messages::{ToolResultMessageContent, UserMessageContent};

    use crate::context::constants::{
        SUMMARIZER_ASSISTANT_TEXT_LIMIT, SUMMARIZER_MAX_SERIALIZED_CHARS,
        SUMMARIZER_TOOL_RESULT_TEXT_LIMIT,
    };

    // -- serialize_messages (delegates to summarizer module, verify integration) --

    #[test]
    fn serialize_empty_messages() {
        let result = serialize_messages(&[]);
        assert!(result.is_empty());
    }

    #[test]
    fn serialize_user_text_message() {
        let msgs = [Message::user("Hello world")];
        let result = serialize_messages(&msgs);
        assert_eq!(result, "[USER] Hello world");
    }

    #[test]
    fn serialize_assistant_text_truncated() {
        let long_text = "a".repeat(500);
        let msgs = [Message::assistant(&long_text)];
        let result = serialize_messages(&msgs);
        assert!(result.contains("[ASSISTANT]"));
        let content = result.strip_prefix("[ASSISTANT] ").unwrap();
        assert_eq!(content.len(), SUMMARIZER_ASSISTANT_TEXT_LIMIT);
    }

    #[test]
    fn serialize_thinking_block() {
        let msgs = [Message::Assistant {
            content: vec![AssistantContent::Thinking {
                thinking: "I need to think about this carefully".into(),
                signature: None,
            }],
            usage: None,
            cost: None,
            stop_reason: None,
            thinking: None,
        }];
        let result = serialize_messages(&msgs);
        assert!(result.starts_with("[THINKING] I need to think"));
    }

    #[test]
    fn serialize_thinking_empty_skipped() {
        let msgs = [Message::Assistant {
            content: vec![AssistantContent::Thinking {
                thinking: String::new(),
                signature: None,
            }],
            usage: None,
            cost: None,
            stop_reason: None,
            thinking: None,
        }];
        let result = serialize_messages(&msgs);
        assert!(result.is_empty());
    }

    #[test]
    fn serialize_tool_call() {
        let mut args = serde_json::Map::new();
        let _ = args.insert("file_path".into(), serde_json::json!("/src/main.rs"));
        let _ = args.insert("command".into(), serde_json::json!("build"));
        let msgs = [Message::Assistant {
            content: vec![AssistantContent::ToolUse {
                id: "tc-1".into(),
                name: "bash".into(),
                arguments: args,
                thought_signature: None,
            }],
            usage: None,
            cost: None,
            stop_reason: None,
            thinking: None,
        }];
        let result = serialize_messages(&msgs);
        assert!(result.starts_with("[TOOL_CALL] bash("));
        assert!(result.contains("file_path: /src/main.rs"));
    }

    #[test]
    fn serialize_tool_result() {
        let msgs = [Message::ToolResult {
            tool_call_id: "tc-1".into(),
            content: ToolResultMessageContent::Text("File content here".into()),
            is_error: None,
        }];
        let result = serialize_messages(&msgs);
        assert!(result.starts_with("[TOOL_RESULT] File content"));
    }

    #[test]
    fn serialize_tool_error() {
        let msgs = [Message::ToolResult {
            tool_call_id: "tc-1".into(),
            content: ToolResultMessageContent::Text("Permission denied".into()),
            is_error: Some(true),
        }];
        let result = serialize_messages(&msgs);
        assert!(result.starts_with("[TOOL_ERROR] Permission denied"));
    }

    #[test]
    fn serialize_tool_result_truncated() {
        let long = "x".repeat(500);
        let msgs = [Message::ToolResult {
            tool_call_id: "tc-1".into(),
            content: ToolResultMessageContent::Text(long),
            is_error: None,
        }];
        let result = serialize_messages(&msgs);
        let content = result.strip_prefix("[TOOL_RESULT] ").unwrap();
        assert_eq!(content.len(), SUMMARIZER_TOOL_RESULT_TEXT_LIMIT);
    }

    #[test]
    fn serialize_truncates_large_output() {
        let big_text = "a".repeat(200);
        let msgs: Vec<Message> = (0..1000).map(|_| Message::user(&big_text)).collect();
        let result = serialize_messages(&msgs);
        assert!(result.len() <= SUMMARIZER_MAX_SERIALIZED_CHARS + 200);
        assert!(result.contains("characters omitted"));
    }

    #[test]
    fn serialize_mixed_conversation() {
        let msgs = vec![
            Message::user("Hello"),
            Message::assistant("Hi there"),
            Message::user("Run a command"),
        ];
        let result = serialize_messages(&msgs);
        let lines: Vec<&str> = result.lines().collect();
        assert_eq!(lines.len(), 3);
        assert!(lines[0].starts_with("[USER]"));
        assert!(lines[1].starts_with("[ASSISTANT]"));
        assert!(lines[2].starts_with("[USER]"));
    }

    // -- parse_summarizer_response (delegates to summarizer module) --

    #[test]
    fn parse_valid_json_response() {
        let raw = r#"{"narrative": "Summary text", "extractedData": {"currentGoal": "Fix bug"}}"#;
        let result = parse_summarizer_response(raw).unwrap();
        assert_eq!(result.narrative, "Summary text");
        assert_eq!(result.extracted_data.current_goal, "Fix bug");
    }

    #[test]
    fn parse_response_with_code_fences() {
        let raw = "```json\n{\"narrative\": \"Summary\"}\n```";
        let result = parse_summarizer_response(raw).unwrap();
        assert_eq!(result.narrative, "Summary");
    }

    #[test]
    fn parse_response_without_extracted_data() {
        let raw = r#"{"narrative": "Just a summary"}"#;
        let result = parse_summarizer_response(raw).unwrap();
        assert_eq!(result.narrative, "Just a summary");
        assert!(result.extracted_data.current_goal.is_empty());
    }

    #[test]
    fn parse_response_empty_narrative_returns_err() {
        let raw = r#"{"narrative": ""}"#;
        assert!(parse_summarizer_response(raw).is_err());
    }

    #[test]
    fn parse_response_missing_narrative_returns_err() {
        let raw = r#"{"summary": "text"}"#;
        assert!(parse_summarizer_response(raw).is_err());
    }

    #[test]
    fn parse_response_invalid_json_returns_err() {
        assert!(parse_summarizer_response("not json at all").is_err());
    }

    #[test]
    fn parse_response_with_extracted_data_fields() {
        let raw = r#"{
            "narrative": "Summary",
            "extractedData": {
                "currentGoal": "Implement auth",
                "completedSteps": ["step 1", "step 2"],
                "pendingTasks": ["task 1"],
                "keyDecisions": [{"decision": "Use JWT", "reason": "Stateless"}],
                "filesModified": ["auth.rs"],
                "topicsDiscussed": ["security"],
                "userPreferences": ["prefer Rust"],
                "importantContext": ["deadline tomorrow"],
                "thinkingInsights": ["complex flow"]
            }
        }"#;
        let result = parse_summarizer_response(raw).unwrap();
        assert_eq!(result.extracted_data.current_goal, "Implement auth");
        assert_eq!(result.extracted_data.completed_steps.len(), 2);
        assert_eq!(result.extracted_data.pending_tasks, vec!["task 1"]);
        assert_eq!(result.extracted_data.key_decisions.len(), 1);
        assert_eq!(result.extracted_data.key_decisions[0].decision, "Use JWT");
        assert_eq!(result.extracted_data.files_modified, vec!["auth.rs"]);
        assert_eq!(result.extracted_data.topics_discussed, vec!["security"]);
    }

    #[test]
    fn parse_response_normalizes_missing_fields() {
        let raw = r#"{"narrative": "Summary", "extractedData": {"currentGoal": "Fix"}}"#;
        let result = parse_summarizer_response(raw).unwrap();
        assert_eq!(result.extracted_data.current_goal, "Fix");
        assert!(result.extracted_data.completed_steps.is_empty());
        assert!(result.extracted_data.key_decisions.is_empty());
    }

    #[test]
    fn parse_response_filters_invalid_decision_entries() {
        let raw = r#"{
            "narrative": "Summary",
            "extractedData": {
                "keyDecisions": [
                    {"decision": "A", "reason": "B"},
                    {"decision": "C"},
                    "not an object"
                ]
            }
        }"#;
        let result = parse_summarizer_response(raw).unwrap();
        assert_eq!(result.extracted_data.key_decisions.len(), 1);
    }

    // -- LlmSummarizer with mock --

    struct MockSpawner {
        result: SubsessionResult,
    }

    #[async_trait]
    impl SubsessionSpawner for MockSpawner {
        async fn spawn_summarizer(&self, _task: &str) -> SubsessionResult {
            self.result.clone()
        }
    }

    #[tokio::test]
    async fn llm_summarizer_success() {
        let spawner = MockSpawner {
            result: SubsessionResult {
                success: true,
                output: Some(r#"{"narrative": "LLM summary"}"#.into()),
                error: None,
            },
        };
        let summarizer = LlmSummarizer::new(spawner);
        let result = summarizer
            .summarize(&[Message::user("test")])
            .await
            .unwrap();
        assert_eq!(result.narrative, "LLM summary");
    }

    #[tokio::test]
    async fn llm_summarizer_fallback_on_failure() {
        let spawner = MockSpawner {
            result: SubsessionResult {
                success: false,
                output: None,
                error: Some("timeout".into()),
            },
        };
        let summarizer = LlmSummarizer::new(spawner);
        let result = summarizer
            .summarize(&[Message::user("test")])
            .await
            .unwrap();
        assert!(result.narrative.contains("1 requests"));
    }

    #[tokio::test]
    async fn llm_summarizer_fallback_on_invalid_json() {
        let spawner = MockSpawner {
            result: SubsessionResult {
                success: true,
                output: Some("not json".into()),
                error: None,
            },
        };
        let summarizer = LlmSummarizer::new(spawner);
        let result = summarizer
            .summarize(&[Message::user("test")])
            .await
            .unwrap();
        assert!(result.narrative.contains("1 requests"));
    }

    #[tokio::test]
    async fn llm_summarizer_fallback_on_no_output() {
        let spawner = MockSpawner {
            result: SubsessionResult {
                success: true,
                output: None,
                error: None,
            },
        };
        let summarizer = LlmSummarizer::new(spawner);
        let result = summarizer
            .summarize(&[Message::user("test")])
            .await
            .unwrap();
        assert!(result.narrative.contains("1 requests"));
    }

    #[tokio::test]
    async fn llm_summarizer_with_code_fences() {
        let spawner = MockSpawner {
            result: SubsessionResult {
                success: true,
                output: Some(
                    "```json\n{\"narrative\": \"Fenced summary\"}\n```".into(),
                ),
                error: None,
            },
        };
        let summarizer = LlmSummarizer::new(spawner);
        let result = summarizer
            .summarize(&[Message::user("test")])
            .await
            .unwrap();
        assert_eq!(result.narrative, "Fenced summary");
    }

    #[tokio::test]
    async fn llm_summarizer_with_extracted_data() {
        let spawner = MockSpawner {
            result: SubsessionResult {
                success: true,
                output: Some(
                    r#"{"narrative": "Summary", "extractedData": {"currentGoal": "Fix auth"}}"#
                        .into(),
                ),
                error: None,
            },
        };
        let summarizer = LlmSummarizer::new(spawner);
        let result = summarizer
            .summarize(&[Message::user("test")])
            .await
            .unwrap();
        assert_eq!(result.narrative, "Summary");
        assert_eq!(result.extracted_data.current_goal, "Fix auth");
    }

    #[tokio::test]
    async fn llm_summarizer_passes_serialized_transcript() {
        use std::sync::{Arc, Mutex};

        struct CapturingSpawner {
            captured: Arc<Mutex<Option<String>>>,
        }

        #[async_trait]
        impl SubsessionSpawner for CapturingSpawner {
            async fn spawn_summarizer(&self, task: &str) -> SubsessionResult {
                *self.captured.lock().unwrap() = Some(task.to_owned());
                SubsessionResult {
                    success: true,
                    output: Some(r#"{"narrative": "ok"}"#.into()),
                    error: None,
                }
            }
        }

        let captured = Arc::new(Mutex::new(None));
        let spawner = CapturingSpawner {
            captured: captured.clone(),
        };
        let summarizer = LlmSummarizer::new(spawner);
        let _ = summarizer
            .summarize(&[Message::user("Hello"), Message::assistant("Hi")])
            .await
            .unwrap();

        let task = captured.lock().unwrap().clone().unwrap();
        assert!(task.contains("[USER] Hello"));
        assert!(task.contains("[ASSISTANT] Hi"));
    }

    // -- UserMessageContent::Blocks (ensure serialize handles blocks) --

    #[test]
    fn serialize_user_with_blocks() {
        let msgs = [Message::User {
            content: UserMessageContent::Blocks(vec![
                tron_core::content::UserContent::Text {
                    text: "First block".into(),
                },
                tron_core::content::UserContent::Text {
                    text: "Second block".into(),
                },
            ]),
            timestamp: None,
        }];
        let result = serialize_messages(&msgs);
        assert!(result.contains("[USER]"));
        assert!(result.contains("First block"));
        assert!(result.contains("Second block"));
    }
}
